Add `Process.clock_gettime` support

alexcwatt opened the pull request:
I am interested in being able to use Timecop with `Process.clock_gettime`. I saw https://github.com/travisjeffery/timecop/issues/220 (the request for this) and https://github.com/travisjeffery/timecop/pull/258 (the PR that implemented it).

This code is nearly the same as https://github.com/travisjeffery/timecop/pull/258; I have only made minor changes.

lavoiesl commented on lib/timecop/time_extensions.rb:
```diff
@@ -158,3 +158,59 @@ def mocked_time_stack_item
+
+      def clock_gettime_mock_time(clock_id, unit = :float_second)
+        mock_time = case clock_id
+                    when Process::CLOCK_MONOTONIC
```
How about `Process::CLOCK_THREAD_CPUTIME_ID`?

alexcwatt commented on test/timecop_test.rb:
```diff
@@ -689,8 +689,93 @@ def test_thread_safe_timecop_returns_after_block
     Timecop.thread_safe = false
   end
 
+  if RUBY_VERSION >= '2.1.0'
```
Yes good point. I think that the assignment `@monotonic = current_monotonic` in `TimeStackItem#initialize` will need to change. I will push a commit to fix when I get a chance.

alexcwatt replied:
Updated to address this.

lavoiesl commented on test/timecop_test.rb:
```diff
@@ -689,8 +689,106 @@ def test_thread_safe_timecop_returns_after_block
+    end
+
+    def test_process_clock_gettime_monotonic_travel
+      current = monotonic
+      Timecop.travel do
+        refute_same monotonic, monotonic, "CLOCK_MONOTONIC is frozen"
+        sleep 0.5
+        assert monotonic > current, "CLOCK_MONOTONIC is not moving forward"
+      end
```
I'm not sure a sleep is necessary, but if it is, I would put it _before_ the travel, to ensure that it does tick forward.

And similar to above, I think 10ms is plenty to make it do at least one tick.
```suggestion
      current = monotonic
      sleep 0.01
      Timecop.travel do
        refute_same monotonic, monotonic, "CLOCK_MONOTONIC is frozen"
        assert monotonic > current, "CLOCK_MONOTONIC is not moving forward"
      end
```

You could define a `TIME_EPSILON = 0.01` constant in the test suite to capture the intent

lavoiesl commented on test/timecop_test.rb:
```diff
@@ -689,8 +689,106 @@ def test_thread_safe_timecop_returns_after_block
+
   private
 
+  if RUBY_VERSION >= '2.1.0'
```
I know this is just following the convention of this codebase, so this isn't a comment on your PR per se:

Ruby 2.1.0 is 11 years old and has been [EOL for 7 years](https://endoflife.date/ruby).

alexcwatt replied:
It would be nice to update the minimum Ruby version in the gemspec; right now it's at 1.9.2.

lavoiesl reviewed (commented):
I haven't tested locally yet, but this LGTM

alexcwatt commented:
@lavoiesl more great feedback - thank you!

joshuacronemeyer commented:
Thanks @lavoiesl and @alexcwatt for making/workshopping this PR (and all the folks over the years that moved the original issues and PRs forward). I am going to have to play around with this a little bit, but don't see why this won't get merged in.

One thought I had was that it would be nice if this new was mentioned in the README. Could you take a crack at that?

alexcwatt commented:
@joshuacronemeyer I made a small change to the README. Let me know what you think. I'm happy to elaborate more but it seems the most important thing is to mention that `Process.clock_gettime` is now supported.

Also let me know if you need me to change anything else, like squashing to a single commit; I kept the commit history originally so reviewers could easily see what I was changing between iterations.

joshuacronemeyer commented on test/timecop_test.rb:
```diff
@@ -689,8 +689,105 @@ def test_thread_safe_timecop_returns_after_block
 
+  if RUBY_VERSION >= '2.1.0'
+    def test_process_clock_gettime_monotonic
+      Timecop.freeze do
```
@alexcwatt just playing around with what you've done. I'm not familiar with Process.clock_gettime but in this test when i add 

```
      refute_same monotonic, monotonic
```
Outside the freeze block the test fails.  If the assert_same line inside the block is to have meaning I think we need to be sure that it can't false positive.

When I change the test like this

```
      t1 = monotonic
      t2 = monotonic
      refute_same t1, t2
      Timecop.freeze do
        t1 = monotonic
        t2 = monotonic
        assert_same t1, t2, "CLOCK_MONOTONIC is not frozen"
      end
```
The result is what you're expecting. I think this is an issue with this test, no? And if that's the case then I maybe some of these other tests need to be looked at carefully. Please have a look.

joshuacronemeyer replied:
Actually that change doesn't matter. If i run it a few times it still fails occasionally. I guess we just need a sleep in between calls to monotonic.

joshuacronemeyer replied:
```
    def test_process_clock_gettime_monotonic
      t1 = monotonic
      sleep(0.1)
      t2 = monotonic
      refute_same t1, t2
      Timecop.freeze do
        t1 = monotonic
        sleep(0.1)
        t2 = monotonic
        assert_same t1, t2, "CLOCK_MONOTONIC is not frozen"
      end
```

lavoiesl replied:
As mentioned at https://github.com/travisjeffery/timecop/pull/419#discussion_r1595497983, I would suggest setting an epsilon constant. We could also add a helper to get two monotonic:

```rb
class TestTimecop < Minitest::Test
  TIME_EPSILON = 0.001 # seconds

  def test_process_clock_gettime_monotonic
    t1, t2 = consecutive_monotonics
    refute_same t1, t2, "CLOCK_MONOTONIC should not be frozen"

    Timecop.freeze do
      t1, t2 = consecutive_monotonics
      assert_same t1, t2, "CLOCK_MONOTONIC should be frozen"
    end
  end

  private
  
  def monotonic
    Process.clock_gettime(Process::CLOCK_MONOTONIC)
  end

  def consecutive_monotonics
    t1 = monotonic
    sleep(TIME_EPSILON)
    t2 = monotonic

    [t1, t2]
  end
end
```

alexcwatt replied:
Thanks for this feedback! I like the suggestion for a fix @lavoiesl.

joshuacronemeyer commented on test/timecop_test.rb:
```diff
@@ -689,8 +689,105 @@ def test_thread_safe_timecop_returns_after_block
+      end
+
+      initial_time = realtime
+      Timecop.freeze(-20) do
```
I don't see any tests that are exercising the behavior of using an unsupported freeze or travel argument like     # 3. Timecop.freeze(date_inst)

Seems like we should test that.

joshuacronemeyer commented on test/timecop_test.rb:
```diff
@@ -689,8 +689,105 @@ def test_thread_safe_timecop_returns_after_block
+      end
+
+      Timecop.travel(-0.5) do
+        refute_same monotonic, monotonic, "CLOCK_MONOTONIC is frozen"
```
This assertion fails occasionally for me. I think it's same problem as above. We need sleep in here.

joshuacronemeyer commented on test/timecop_test.rb:
```diff
@@ -689,8 +689,105 @@ def test_thread_safe_timecop_returns_after_block
 
+  if RUBY_VERSION >= '2.1.0'
+    def monotonic
+      Process.clock_gettime(Process::CLOCK_MONOTONIC)
```
We're always using the default nanosecond unit for all our tests. I think it would be nice to have a test where we freeze and then make sure that Process.clock_gettime is converting the units correctly for seconds, milliseconds, etc.

joshuacronemeyer reviewed (changes requested):
Seems like the code works. I've flagged some problems with the tests and some additional tests i would like. Thanks again!

alexcwatt commented:
@joshuacronemeyer Thank you for the detailed review! I obviously missed some things. I think this version addresses your comments. Please let me know what you think when you have another look. I meant to address this feedback sooner but couldn't get to it last weekend.

I've preserved commit history so you can see what changed, but when this is ready to be merged, we can squash everything.

alexcwatt commented on test/timecop_with_process_clock_test.rb:
```diff
@@ -0,0 +1,168 @@
+require 'timecop'
+
+class TestTimecopWithProcessClock < Minitest::Test
+  TIME_EPSILON = 0.001 # seconds - represents enough time for Process.clock_gettime to have advanced if not frozen
```
Open to other ideas for this constant name to better represent the meaning that I'm currently explaining with the comment.

joshuacronemeyer reviewed (approved):
Looking good. Thanks.

nerdrew commented:
Can we make this stub opt-in? It seems to break various low-level use cases when it's enabled. e.g. https://github.com/travisjeffery/timecop/issues/425 and the fluentd PR above...
